gma/src/android/native_ad_internal_android.cc

```diff
@@ -138,6 +138,133 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
+    const Variant& impression_data) {
+  firebase::MutexLock lock(mutex_);
+
+  if (!initialized_) {
+    return CreateAndCompleteFuture(kNativeAdFnRecordImpression,
+                                   kAdErrorCodeUninitialized,
+                                   kAdUninitializedErrorMessage, &future_data_);
+  }
+
+  jobject impression_bundle = variantmap_to_bundle(impression_data);
+
+  if (impression_bundle == nullptr) {
+    return CreateAndCompleteFuture(
+        kNativeAdFnRecordImpression, kAdErrorCodeInvalidArgument,
+        kUnsupportedVariantTypeErrorMessage, &future_data_);
+  }
+
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  FutureCallbackData<void>* callback_data =
+      CreateVoidFutureCallbackData(kNativeAdFnRecordImpression, &future_data_);
+  Future<void> future =
+      MakeFuture(&future_data_.future_impl, callback_data->future_handle);
+  env->CallVoidMethod(
+      helper_,
+      native_ad_helper::GetMethodId(native_ad_helper::kRecordImpression),
+      reinterpret_cast<jlong>(callback_data), impression_bundle);
+
+  util::CheckAndClearJniExceptions(env);
+
+  return future;
+}
+
+Future<void> NativeAdInternalAndroid::PerformClick(const Variant& click_data) {
+  firebase::MutexLock lock(mutex_);
+
+  if (!initialized_) {
+    return CreateAndCompleteFuture(kNativeAdFnPerformClick,
+                                   kAdErrorCodeUninitialized,
+                                   kAdUninitializedErrorMessage, &future_data_);
+  }
+
+  jobject click_bundle = variantmap_to_bundle(click_data);
+
+  if (click_bundle == nullptr) {
+    return CreateAndCompleteFuture(
+        kNativeAdFnPerformClick, kAdErrorCodeInvalidArgument,
+        kUnsupportedVariantTypeErrorMessage, &future_data_);
+  }
+
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  FutureCallbackData<void>* callback_data =
+      CreateVoidFutureCallbackData(kNativeAdFnPerformClick, &future_data_);
+  Future<void> future =
+      MakeFuture(&future_data_.future_impl, callback_data->future_handle);
+  env->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kPerformClick),
+      reinterpret_cast<jlong>(callback_data), click_bundle);
+
+  util::CheckAndClearJniExceptions(env);
+
+  return future;
+}
+
+jobject NativeAdInternalAndroid::variantmap_to_bundle(
+    const Variant& variant_data) {
+  if (!variant_data.is_map()) {
+    return nullptr;
+  }
+
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  jobject variant_bundle =
+      env->NewObject(util::bundle::GetClass(),
+                     util::bundle::GetMethodId(util::bundle::kConstructor));
+  FIREBASE_ASSERT(variant_bundle);
+
+  for (const auto& kvp : variant_data.map()) {
+    const Variant& key = kvp.first;
+    const Variant& value = kvp.second;
+
+    if (!key.is_string()) {
+      return nullptr;
+    }
+    jstring key_str = env->NewStringUTF(key.string_value());
+
```

Comment:
Might want to add `util::CheckAndClearJniExceptions(env);` here.

Reply:
Done.